# Capstone Project - The Battle of the Neighborhoods 
### Applied Data Science Capstone by IBM/Coursera

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Introduction: Business Problem <a name="introduction"></a>

In this project we will try to find an optimal location for a restaurant. Specifically, this report will be targeted to stakeholders interested in opening a Vietnamese restaurant in Atlanta, GA,USA.

Since there are lots of restaurants in Atlanta we will try to detect locations that are not already crowded with restaurants. We are also particularly interested in areas with no Vietnamese restaurants in vicinity. We would also prefer locations as close to city center as possible, assuming that first two conditions are met.

We will use our data science powers to generate a few most promising neighborhoods based on these criteria. Advantages of each area will then be clearly expressed so that best possible final location can be chosen by stakeholders.

## Data <a name="data"></a>

Based on definition of our problem, factors that will influence our decision are:
* Number of existing restaurants in the neighborhood (any type of restaurant)
* Number of and distance to Vietnamese restaurants in the neighborhood, if any
* Distance of neighborhood from city center

We decided to use regularly spaced grid of locations, centered around city center, to define our neighborhoods.

Following data sources will be needed to extract/generate the required information:
* Centers of candidate areas will be generated algorithmically and approximate addresses of centers of those areas will be obtained using **Google Maps API reverse geocoding**
* Cumber of restaurants and their type and location in every neighborhood will be obtained using **Foursquare API**
* Coordinate of Atlanta center will be obtained using **Google Maps API geocoding** of well known Atlanta location (Downtown)

### Neighborhood Candidates

Let's create latitude & longitude coordinates for centroids of our candidate neighborhoods. We will create a grid of cells covering our area of interest which is aprox. 12x12 killometers centered around Atlanta city center.

Let's first find the latitude & longitude of Atlanta city center, using specific, well known address and Google Maps geocoding API.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
# uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
# uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning) 
print('Libraries imported.')

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\daile\Anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1h             |       he774522_0         5.8 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         5.8 MB

The following NEW packages will be INSTALLED:

  python_abi         conda-forge/win-64::python_abi-3.8-1_cp38

The following packages will be UPDATED:

  ca-certificates    pkgs/main::ca-certificates-2020.10.14~ --> conda-forge::ca-certificates-2020.11.8-h5b45459_0
  certifi            pkgs/main/noarch::certifi-2020.6.20-p~ --> conda-forge/win-64::certifi-2020.11.8-py38haa244fe_0

The following packages will be SUPERSEDED by a higher-priority channel:

  conda               pkgs/main::conda-4

In [2]:
#import geodata
import requests 
url='https://opendata.arcgis.com/datasets/297d3d69d8ab4c6ba5f9264ad5e75c0a_3.geojson'
r = requests.get(url, allow_redirects=True)
open('atlanta.json', 'wb').write(r.content)


732049

In [3]:
with open('atlanta.json') as json_data:
    atlanta_data = json.load(json_data)
atlanta_data

{'type': 'FeatureCollection',
 'name': 'Neighborhoods',
 'crs': {'type': 'name',
  'properties': {'name': 'urn:ogc:def:crs:OGC:1.3:CRS84'}},
 'features': [{'type': 'Feature',
   'properties': {'OBJECTID': 148,
    'LOCALID': None,
    'NAME': 'Atlanta University Center',
    'GEOTYPE': 'Neighborhood',
    'FULLFIPS': None,
    'LEGALAREA': None,
    'EFFECTDATE': '1970/01/01 00:00:00+00',
    'ENDDATE': '1970/01/01 00:00:00+00',
    'SRCREF': None,
    'ACRES': 331.57,
    'SQMILES': 0.52,
    'OLDNAME': 'Atlanta University',
    'NPU': 'T',
    'CREATED_USER': None,
    'CREATED_DATE': '1970/01/01 00:00:00+00',
    'LAST_EDITED_USER': 'GIS',
    'LAST_EDITED_DATE': '2018/05/15 13:24:10+00',
    'GLOBALID': '{D6CAF694-B83F-4CB9-B351-CA3E3145D0C7}'},
   'geometry': {'type': 'Polygon',
    'coordinates': [[[-84.40279850020488, 33.75464936754819],
      [-84.40418541506223, 33.75469273994923],
      [-84.41277106114923, 33.75469893978386],
      [-84.41756016330804, 33.7547357616783],
   

In [4]:
atl_df=pd.DataFrame
atlanta_data['features']

[{'type': 'Feature',
  'properties': {'OBJECTID': 148,
   'LOCALID': None,
   'NAME': 'Atlanta University Center',
   'GEOTYPE': 'Neighborhood',
   'FULLFIPS': None,
   'LEGALAREA': None,
   'EFFECTDATE': '1970/01/01 00:00:00+00',
   'ENDDATE': '1970/01/01 00:00:00+00',
   'SRCREF': None,
   'ACRES': 331.57,
   'SQMILES': 0.52,
   'OLDNAME': 'Atlanta University',
   'NPU': 'T',
   'CREATED_USER': None,
   'CREATED_DATE': '1970/01/01 00:00:00+00',
   'LAST_EDITED_USER': 'GIS',
   'LAST_EDITED_DATE': '2018/05/15 13:24:10+00',
   'GLOBALID': '{D6CAF694-B83F-4CB9-B351-CA3E3145D0C7}'},
  'geometry': {'type': 'Polygon',
   'coordinates': [[[-84.40279850020488, 33.75464936754819],
     [-84.40418541506223, 33.75469273994923],
     [-84.41277106114923, 33.75469893978386],
     [-84.41756016330804, 33.7547357616783],
     [-84.41753241461095, 33.749179980814375],
     [-84.41754771430935, 33.74544352521355],
     [-84.41754575009267, 33.74448645056426],
     [-84.41534808934827, 33.744449672859

In [5]:
address = 'Atlanta City, GA'

geolocator = Nominatim(user_agent="Atl_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

print('The geograpical coordinate of Atlanta City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Atlanta City are 33.7490987, -84.3901849.


Now let's create a grid of area candidates, equaly spaced, centered around city center and within ~6km from Downtown. Our neighborhoods will be defined as circular areas with a radius of 300 meters, so our neighborhood centers will be 600 meters apart.

To accurately calculate distances we need to create our grid of locations in Cartesian 2D coordinate system which allows us to calculate distances in meters (not in latitude/longitude degrees). Then we'll project those coordinates back to latitude/longitude degrees to be shown on Folium map. So let's create functions to convert between WGS84 spherical coordinate system (latitude/longitude degrees) and UTM Cartesian coordinate system (X/Y coordinates in  meters).

In [6]:
#!pip install shapely
import shapely.geometry

#!pip install pyproj
import pyproj

import math

def lonlat_to_xy(lon, lat):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    xy = pyproj.transform(proj_latlon, proj_xy, lon, lat)
    return xy[0], xy[1]

def xy_to_lonlat(x, y):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    lonlat = pyproj.transform(proj_xy, proj_latlon, x, y)
    return lonlat[0], lonlat[1]

def calc_xy_distance(x1, y1, x2, y2):
    dx = x2 - x1
    dy = y2 - y1
    return math.sqrt(dx*dx + dy*dy)

print('Coordinate transformation check')
print('-------------------------------')
print('Atlanta center longitude={}, latitude={}'.format(longitude,latitude))
x, y = lonlat_to_xy(longitude, latitude)
print('Atlanta center UTM X={}, Y={}'.format(x, y))
lo, la = xy_to_lonlat(x, y)
print('Atlanta center longitude={}, latitude={}'.format(lo, la))

Coordinate transformation check
-------------------------------
Atlanta center longitude=-84.3901849, latitude=33.7490987
Atlanta center UTM X=-6878875.466919403, Y=11519684.702060048
Atlanta center longitude=-84.39018490000197, latitude=33.749098700009355


Let's create a **hexagonal grid of cells**: we offset every other row, and adjust vertical row spacing so that **every cell center is equally distant from all it's neighbors**.

In [26]:
atl_center_x, atl_center_y = lonlat_to_xy(longitude, latitude) # City center in Cartesian coordinates

k = math.sqrt(3) / 2 # Vertical offset for hexagonal grid cells
x_min = atl_center_x - 6000
x_step = 600
y_min = atl_center_y - 6000 - (int(21/k)*k*600 - 12000)/2
y_step = 600 * k 

latitudes = []
longitudes = []
distances_from_center = []
xs = []
ys = []
for i in range(0, int(21/k)):
    y = y_min + i * y_step
    x_offset = 300 if i%2==0 else 0
    for j in range(0, 21):
        x = x_min + j * x_step + x_offset
        distance_from_center = calc_xy_distance(atl_center_x, atl_center_y, x, y)
        if (distance_from_center <= 6001):
            lon, lat = xy_to_lonlat(x, y)
            latitudes.append(lat)
            longitudes.append(lon)
            distances_from_center.append(distance_from_center)
            xs.append(x)
            ys.append(y)

print(len(latitudes), 'candidate neighborhood centers generated.')

364 candidate neighborhood centers generated.


Let's visualize the data we have so far: city center location and candidate neighborhood centers:

In [25]:
import folium
# create map of Atl using latitude and longitude values
map_atl= folium.Map(location=[latitude, longitude], zoom_start=13)
folium.Marker([latitude, longitude], popup='Atlanta').add_to(map_atl)
for lat, lon in zip(latitudes, longitudes):
    folium.Circle([lat, lon], radius=300, color='red', fill=False).add_to(map_atl)  
map_atl.save('C:/users/daile/Dropbox/R Python/IBM capstone project/Fig1.html')


OK, we now have the coordinates of centers of neighborhoods/areas to be evaluated, equally spaced (distance from every point to it's neighbors is exactly the same) and within ~6km from Downtown. 

Let's now use Google Maps API to get approximate addresses of those locations.

In [27]:
google_api_key='AIzaSyDhYYiEvsDxhKeU9jXM69XniQhNct1x5zU'
def get_address(api_key, lat, long, verbose=False):
    try:
        url = 'https://maps.googleapis.com/maps/api/geocode/json?key={}&latlng={},{}'.format(api_key, lat, long)
        response = requests.get(url).json()
        if verbose:
            print('Google Maps API JSON result =>', response)
        results = response['results']
        address = results[0]['formatted_address']
        return address
    except:
        return None

addr = get_address(google_api_key, latitude, longitude)
print('Reverse geocoding check')
print('-----------------------')
print('Address of [{}, {}] is: {}'.format(latitude, longitude, addr))

Reverse geocoding check
-----------------------
Address of [33.7490987, -84.3901849] is: 68 Mitchell St SW, Atlanta, GA 30303, USA


In [28]:
print('Obtaining location addresses: ', end='')
addresses = []
for lat, lon in zip(latitudes, longitudes):
    address = get_address(google_api_key, lat, lon)
    if address is None:
        address = 'NO ADDRESS'
    address = address.replace(', USA', '') 
    addresses.append(address)
    print(' .', end='')
print(' done.')

Obtaining location addresses:  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . done.


In [29]:
addresses[150:170]

['Cloudman Residence Hall, 661 Techwood Dr NW, Atlanta, GA 30332',
 '175 4th St NW, Atlanta, GA 30313',
 '237 Cherry St NW, Atlanta, GA 30318',
 '360 Hill St SE, Atlanta, GA 30312',
 '1043 Hill St SE, Atlanta, GA 30315',
 '222 Tuskegee St SE, Atlanta, GA 30315',
 '103 South Ave SE, Atlanta, GA 30315',
 'Bass Street, Atlanta, GA 30315',
 '142 Georgia Ave SE, Atlanta, GA 30312',
 '500 Martin St SE, Atlanta, GA 30312',
 '101 Rawson St SE, Atlanta, GA 30312',
 'Memorial Dr SE @ Fraser St SE, Atlanta, GA 30312',
 'Ga State Station Piedmont Ave, Atlanta, GA 30334',
 'Wall St at Central Ave, Atlanta, GA 30303',
 '67 Broad St NW, Atlanta, GA 30303',
 '108 Ted Turner Dr NW, Atlanta, GA 30303',
 '262 Williams St NW, Atlanta, GA 30313',
 '165 Ivan Allen Jr Blvd NW, Atlanta, GA 30308',
 '202 Pine St NW, Atlanta, GA 30313',
 '230 North Avenue NW, Atlanta, GA 30313']

In [30]:
#Save these addresses to pkl file
import pandas as pd

df_locations = pd.DataFrame({'Address': addresses,
                             'Latitude': latitudes,
                             'Longitude': longitudes,
                             'X': xs,
                             'Y': ys,
                             'Distance from center': distances_from_center})

df_locations.head(10)

Address   Latitude  Longitude  \
0          197 Wilbur Ave SE, Atlanta, GA 30316  33.748513 -84.353047   
1         81 Weatherby St SE, Atlanta, GA 30316  33.751494 -84.354095   
2            1004 Hanover Ct, Atlanta, GA 30307  33.754474 -84.355143   
3        124 Elizabeth St NE, Atlanta, GA 30307  33.757455 -84.356191   
4         926 Waverly Way NE, Atlanta, GA 30307  33.760437 -84.357240   
5        325 Elizabeth St NE, Atlanta, GA 30307  33.763418 -84.358288   
6           400 Village Pkwy, Atlanta, GA 30306  33.766400 -84.359337   
7  982 N Ormewood Park Dr SE, Atlanta, GA 30316  33.743284 -84.354565   
8       968 Memorial Walk SE, Atlanta, GA 30316  33.746265 -84.355613   
9          934 Mauldin St SE, Atlanta, GA 30316  33.749245 -84.356661   

              X             Y  Distance from center  
0 -6.880675e+06  1.151397e+07           5992.495307  
1 -6.880075e+06  1.151397e+07           5840.376700  
2 -6.879475e+06  1.151397e+07           5747.173218  
3 -6.878875e+06  1.151397e+07           5715.767665  
4 -6.878275e+06  1.151397e+07           5747.173218  
5 -6.877675e+06  1.151397e+07           5840.376700  
6 -6.877075e+06  1.151397e+07           5992.495307  
7 -6.881575e+06  1.151449e+07           5855.766389  
8 -6.880975e+06  1.151449e+07           5604.462508  
9 -6.880375e+06  1.151449e+07           5408.326913

In [ ]:
df_locations.to_pickle('./locations.pkl')    

### Foursquare
Now that we have our location candidates, let's use Foursquare API to get info on restaurants in each neighborhood.

We're interested in venues in 'food' category, but only those that are proper restaurants - coffe shops, pizza places, bakeries etc. are not direct competitors so we don't care about those. So we will include in out list only venues that have 'restaurant' in category name, and we'll make sure to detect and include all the subcategories of specific 'Vietnamese restaurant' category, as we need info on Vietnamese restaurants in the neighborhood.

In [31]:
version = '20180724'
client_id = '15VRGSNSBDDGBC0CF25VAVY3NKORSIPQKTUZCE4GEGVHXU55' 
client_secret = 'BZCEVH02DBGMXPXYJQ4RS32K55FDDQRBBCXVWSQFAVJHNQXC' 
food_category = '4d4b7105d754a06374d81259'
asian_restaurant_category='4bf58dd8d48988d142941735'
vietnamese_restaurant_category ='4bf58dd8d48988d14a941735'
def is_restaurant(categories, specific_filter=None):
    restaurant_words = ['restaurant', 'diner', 'taverna', 'steakhouse']
    restaurant = False
    specific = False
    for c in categories:
        category_name = c[0].lower()
        category_id = c[1]
        for r in restaurant_words:
            if r in category_name:
                restaurant = True
        if 'fast food' in category_name:
            restaurant = False
        if not(specific_filter is None) and (category_id in specific_filter):
            specific = True
            restaurant = True
    return restaurant, specific

def get_categories(categories):
    return [(cat['name'], cat['id']) for cat in categories]

def format_address(location):
    address = ', '.join(location['formattedAddress'])
    address = address.replace(', USA', '')
    return address

def get_venues_near_location(lat, lon, category, client_id, client_secret, radius=500, limit=100):
    version = '20180724'
    url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&categoryId={}&radius={}&limit={}'.format(
        client_id, client_secret, version, lat, lon, category, radius, limit)
    try:
        results = requests.get(url).json()['response']['groups'][0]['items']
        venues = [(item['venue']['id'],
                   item['venue']['name'],
                   get_categories(item['venue']['categories']),
                   (item['venue']['location']['lat'], item['venue']['location']['lng']),
                   format_address(item['venue']['location']),
                   item['venue']['location']['distance']) for item in results]        
    except:
        venues = []
    return venues

In [32]:
# Let's now go over our neighborhood locations and get nearby restaurants; we'll also maintain a dictionary of all found restaurants and all found Vientamese restaurants

import pickle

def get_restaurants(lats, lons):
    restaurants = {}
    vietnamese_restaurants = {}
    location_restaurants = []

    print('Obtaining venues around candidate locations:', end='')
    for lat, lon in zip(lats, lons):
        # Using radius=350 to meke sure we have overlaps/full coverage so we don't miss any restaurant (we're using dictionaries to remove any duplicates resulting from area overlaps)
        venues = get_venues_near_location(lat, lon, food_category, client_id, client_secret, radius=350, limit=100)
        area_restaurants = []
        for venue in venues:
            venue_id = venue[0]
            venue_name = venue[1]
            venue_categories = venue[2]
            venue_latlon = venue[3]
            venue_address = venue[4]
            venue_distance = venue[5]
            is_res, is_vietnamese = is_restaurant(venue_categories, specific_filter=vietnamese_restaurant_category)
            if is_res:
                x, y = lonlat_to_xy(venue_latlon[1], venue_latlon[0])
                restaurant = (venue_id, venue_name, venue_latlon[0], venue_latlon[1], venue_address, venue_distance, is_vietnamese, x, y)
                if venue_distance<=300:
                    area_restaurants.append(restaurant)
                restaurants[venue_id] = restaurant
                if is_vietnamese:
                    vietnamese_restaurants[venue_id] = restaurant
        location_restaurants.append(area_restaurants)
        print(' .', end='')
    print(' done.')
    return restaurants, vietnamese_restaurants, location_restaurants

# Try to load from local file system in case we did this before
restaurants = {}
vietnamese_restaurants = {}
location_restaurants = []
loaded = False
try:
    with open('restaurants_350.pkl', 'rb') as f:
        restaurants = pickle.load(f)
    with open('vietnamese_restaurants_350.pkl', 'rb') as f:
        vietnamese_restaurants = pickle.load(f)
    with open('location_restaurants_350.pkl', 'rb') as f:
        location_restaurants = pickle.load(f)
    print('Restaurant data loaded.')
    loaded = True
except:
    pass

# If load failed use the Foursquare API to get the data
if not loaded:
    restaurants, vietnamese_restaurants, location_restaurants = get_restaurants(latitudes, longitudes)
    
    # Let's persists this in local file system
    with open('restaurants_350.pkl', 'wb') as f:
        pickle.dump(restaurants, f)
    with open('vietnamese_restaurants_350.pkl', 'wb') as f:
        pickle.dump(vietnamese_restaurants, f)
    with open('location_restaurants_350.pkl', 'wb') as f:
        pickle.dump(location_restaurants, f)

Restaurant data loaded.


In [33]:
import numpy as np

print('Total number of restaurants:', len(restaurants))
print('Total number of Vietnamese restaurants:', len(vietnamese_restaurants))
print('Percentage of Vietnamese restaurants: {:.2f}%'.format(len(vietnamese_restaurants) / len(restaurants) * 100))
print('Average number of restaurants in neighborhood:', np.array([len(r) for r in location_restaurants]).mean())

Total number of restaurants: 375
Total number of Vietnamese restaurants: 11
Percentage of Vietnamese restaurants: 2.93%
Average number of restaurants in neighborhood: 2.3736263736263736


In [34]:
print('List of all restaurants')
print('-----------------------')
for r in list(restaurants.values())[:10]:
    print(r)
print('...')
print('Total:', len(restaurants))

List of all restaurants
-----------------------
('4c12a44fb7b9c928c82ba837', 'Home Grown', 33.74674699478201, -84.35594830537276, '968 Memorial Dr SE (at Gibson St SE), Atlanta, GA 30316, United States', 304, False, -6880870.798927291, 11514514.206224235)
('51d73636498ea9799c37b956', 'dead animals on the wall restaurant', 33.752496115876625, -84.35473123700214, 'Edgewood Street, Atlanta, GA, United States', 126, False, -6879860.768529584, 11514012.840909448)
('4dbc5771cda109aa6c98f09c', 'One Eared Stag', 33.757232712502784, -84.354153829226, '1029 Edgewood Ave NE (Hurt St), Atlanta, GA 30307, United States', 190, False, -6879009.359599424, 11513665.712614281)
('5c09ad04c36588002c1e0cfa', "Lloyd's Restaurant and Lounge", 33.75488784098578, -84.35757637023926, '900 DeKalb Ave, NE Atlanta, GA 30307, United States', 114, False, -6879288.241716245, 11514323.01656757)
('56746b66498e1c81cc3a3dbc', 'Bread  & Butterfly', 33.76246580642556, -84.35804544032095, '290 Elizabeth St, Atlanta, GA 3030

In [35]:
print('List of Vietnamese restaurants')
print('---------------------------')
for r in list(vietnamese_restaurants.values())[:10]:
    print(r)
print('...')
print('Total:', len(vietnamese_restaurants))

List of Vietnamese restaurants
---------------------------
('5ba19d19724750002ce25df4', 'Pho Nam Ksm', 33.757214, -84.36399, '99 Krog St NE, Atlanta, GA 30307, United States', 194, True, -6878565.5801921915, 11515188.894139033)
('4e2463f7a80968cf3e5cd41b', 'Vietnemese Noodle', 33.764076, -84.374183, 'Atlanta, GA, United States', 105, True, -6876830.7115759, 11516392.031408517)
('58b60fea668af7464eb7a035', 'Hungry Ghost', 33.754199, -84.376299, '345 Edgewood Ave SE, Atlanta, GA 30312, United States', 200, True, -6878563.780383785, 11517257.834719563)
('59efd93b59c4231afeb41e8d', 'Anh’s Kitchen', 33.777942, -84.384174, '855 Peachtree St NE Ste 2, Atlanta, GA 30308, United States', 244, True, -6873808.466951068, 11517181.230887452)
('58ae310b084be53884dcbe4f', 'Pho King', 33.78042975591016, -84.38745879910743, 'Atlanta, GA 30309, United States', 334, True, -6873198.409154813, 11517553.363799166)
('58adda80964d7851965e1ce6', 'Pho King - Midtown', 33.779143, -84.38733, '950 W Peachtree St N

In [36]:
print('Restaurants around location')
print('---------------------------')
for i in range(100, 110):
    rs = location_restaurants[i][:8]
    names = ', '.join([r[1] for r in rs])
    print('Restaurants around location {}: {}'.format(i+1, names))

Restaurants around location
---------------------------
Restaurants around location 101: Dumpling Girl
Restaurants around location 102: Grant Central Annex
Restaurants around location 103: Mezcalito's Cocina & Tequila Bar, Six Feet Under Pub & Fish House, Republic Social House
Restaurants around location 104: Mezcalito's Cocina & Tequila Bar, Bahama Bay Conch Bar/ Grille
Restaurants around location 105: Budi's Sushi, American Roadhouse, Bahama Bay Conch Bar/ Grille
Restaurants around location 106: Arepa Mia, Sweet Auburn Seafood, Metro Deli and Soul Food, The Original Jamaican Restaurant, BQE Restaurant, Afrodish, Delightful EATZ, Hungry Ghost
Restaurants around location 107: Sweet Auburn Seafood, Arepa Mia, Mangos Caribbean Restaurant, Metro Deli and Soul Food, BQE Restaurant, Kenley's Catering & Restaurant, Afrodish, Wok n roll
Restaurants around location 108: Fandangles
Restaurants around location 109: Nikolai's Roof, Malice's Restaurant, Point Of View, Marketplace
Restaurants aroun

Let's now see all the collected restaurants in our area of interest on map, and let's also show Vietnames restaurants in different color.

In [37]:
map_atl = folium.Map(location=[latitude, longitude], zoom_start=13)
folium.Marker([latitude, longitude], popup='Atlanta').add_to(map_atl)
for res in restaurants.values():
    lat = res[2]; lon = res[3]
    is_vietnamese = res[6]
    color = 'red' if is_vietnamese else 'blue'
    folium.CircleMarker([lat, lon], radius=3, color=color, fill=True, fill_color=color, fill_opacity=1).add_to(map_atl)
map_atl

This concludes the data gathering phase - we're now ready to use this data for analysis to produce the report on optimal locations for a new Vietnamese restaurant!

## Methodology <a name="methodology"></a>

In this project we will direct our efforts on detecting areas of Atlanta that have low restaurant density, particularly those with low number of Vietnamese restaurants. We will limit our analysis to area ~6km around city center.

In first step we have collected the required **data: location and type (category) of every restaurant within 6km from Atlanta center** . We have also **identified Vietnamese restaurants** (according to Foursquare categorization).

Second step in our analysis will be calculation and exploration of '**restaurant density**' across different areas of Atlanta - we will use **heatmaps** to identify a few promising areas close to center with low number of restaurants in general (*and* no Vietnamese restaurants in vicinity) and focus our attention on those areas.

In third and final step we will focus on most promising areas and within those create **clusters of locations that meet some basic requirements** established in discussion with stakeholders: we will take into consideration locations with **no more than two restaurants in radius of 250 meters**, and we want locations **without Vietnamese restaurants in radius of 400 meters**. We will present map of all such locations but also create clusters (using **k-means clustering**) of those locations to identify general zones / neighborhoods / addresses which should be a starting point for final 'street level' exploration and search for optimal venue location by stakeholders.

## Analysis <a name="analysis"></a>

Let's perform some basic explanatory data analysis and derive some additional info from our raw data. First let's count the **number of restaurants in every area candidate**:

In [38]:
location_restaurants_count = [len(res) for res in location_restaurants]

df_locations['Restaurants in area'] = location_restaurants_count

print('Average number of restaurants in every area with radius=300m:', np.array(location_restaurants_count).mean())

df_locations.head(10)

Average number of restaurants in every area with radius=300m: 2.3736263736263736


Address   Latitude  Longitude  \
0          197 Wilbur Ave SE, Atlanta, GA 30316  33.748513 -84.353047   
1         81 Weatherby St SE, Atlanta, GA 30316  33.751494 -84.354095   
2            1004 Hanover Ct, Atlanta, GA 30307  33.754474 -84.355143   
3        124 Elizabeth St NE, Atlanta, GA 30307  33.757455 -84.356191   
4         926 Waverly Way NE, Atlanta, GA 30307  33.760437 -84.357240   
5        325 Elizabeth St NE, Atlanta, GA 30307  33.763418 -84.358288   
6           400 Village Pkwy, Atlanta, GA 30306  33.766400 -84.359337   
7  982 N Ormewood Park Dr SE, Atlanta, GA 30316  33.743284 -84.354565   
8       968 Memorial Walk SE, Atlanta, GA 30316  33.746265 -84.355613   
9          934 Mauldin St SE, Atlanta, GA 30316  33.749245 -84.356661   

              X             Y  Distance from center  Restaurants in area  
0 -6.880675e+06  1.151397e+07           5992.495307                    0  
1 -6.880075e+06  1.151397e+07           5840.376700                    1  
2 -6.879475e+06  1.151397e+07           5747.173218                    1  
3 -6.878875e+06  1.151397e+07           5715.767665                    1  
4 -6.878275e+06  1.151397e+07           5747.173218                    8  
5 -6.877675e+06  1.151397e+07           5840.376700                   11  
6 -6.877075e+06  1.151397e+07           5992.495307                    0  
7 -6.881575e+06  1.151449e+07           5855.766389                    1  
8 -6.880975e+06  1.151449e+07           5604.462508                    3  
9 -6.880375e+06  1.151449e+07           5408.326913                    4

OK, now let's calculate the **distance to nearest Vietnamese restaurant from every area candidate center** (not only those within 300m - we want distance to closest one, regardless of how distant it is).

In [39]:
distances_to_vietnamese_restaurant = []

for area_x, area_y in zip(xs, ys):
    min_distance = 10000
    for res in vietnamese_restaurants.values():
        res_x = res[7]
        res_y = res[8]
        d = calc_xy_distance(area_x, area_y, res_x, res_y)
        if d<min_distance:
            min_distance = d
    distances_to_vietnamese_restaurant.append(min_distance)

df_locations['Distance to Vietnamese restaurant'] = distances_to_vietnamese_restaurant

In [40]:
df_locations.head(10)

Address   Latitude  Longitude  \
0          197 Wilbur Ave SE, Atlanta, GA 30316  33.748513 -84.353047   
1         81 Weatherby St SE, Atlanta, GA 30316  33.751494 -84.354095   
2            1004 Hanover Ct, Atlanta, GA 30307  33.754474 -84.355143   
3        124 Elizabeth St NE, Atlanta, GA 30307  33.757455 -84.356191   
4         926 Waverly Way NE, Atlanta, GA 30307  33.760437 -84.357240   
5        325 Elizabeth St NE, Atlanta, GA 30307  33.763418 -84.358288   
6           400 Village Pkwy, Atlanta, GA 30306  33.766400 -84.359337   
7  982 N Ormewood Park Dr SE, Atlanta, GA 30316  33.743284 -84.354565   
8       968 Memorial Walk SE, Atlanta, GA 30316  33.746265 -84.355613   
9          934 Mauldin St SE, Atlanta, GA 30316  33.749245 -84.356661   

              X             Y  Distance from center  Restaurants in area  \
0 -6.880675e+06  1.151397e+07           5992.495307                    0   
1 -6.880075e+06  1.151397e+07           5840.376700                    1   
2 -6.879475e+06  1.151397e+07           5747.173218                    1   
3 -6.878875e+06  1.151397e+07           5715.767665                    1   
4 -6.878275e+06  1.151397e+07           5747.173218                    8   
5 -6.877675e+06  1.151397e+07           5840.376700                   11   
6 -6.877075e+06  1.151397e+07           5992.495307                    0   
7 -6.881575e+06  1.151449e+07           5855.766389                    1   
8 -6.880975e+06  1.151449e+07           5604.462508                    3   
9 -6.880375e+06  1.151449e+07           5408.326913                    4   

   Distance to Vietnamese restaurant  
0                        2437.195884  
1                        1941.149068  
2                        1521.905264  
3                        1258.702332  
4                        1253.980657  
5                        1510.166685  
6                        1925.808750  
7                        3090.291334  
8                        2509.588902  
9                        1940.662873

In [41]:
print('Average distance to closest Vietnamese restaurant from each area center:', df_locations['Distance to Vietnamese restaurant'].mean())

Average distance to closest Vietnamese restaurant from each area center: 3092.342144383127


OK, so **on average Vietnamese restaurant can be found within ~3.1km** from every area center candidate. That's fairly far.

Let's crete a map showing **heatmap / density of restaurants** and try to extract some meaningfull info from that. Also, let's show **borders of Atlanta neighborhoods** on our map and a few circles indicating distance of 1km, 2km and 3km from Atlanta center.

In [42]:
def boroughs_style(feature):
    return { 'color': 'blue', 'fill': False }

In [43]:
restaurant_latlons = [[res[2], res[3]] for res in restaurants.values()]

vietnamese_latlons = [[res[2], res[3]] for res in vietnamese_restaurants.values()]

In [45]:
from folium import plugins
from folium.plugins import HeatMap

map_atl = folium.Map(location=[latitude, longitude], zoom_start=13)
folium.TileLayer('cartodbpositron').add_to(map_atl) #cartodbpositron cartodbdark_matter
HeatMap(restaurant_latlons).add_to(map_atl)
folium.Marker([latitude, longitude]).add_to(map_atl)
folium.Circle([latitude, longitude], radius=1000, fill=False, color='white').add_to(map_atl)
folium.Circle([latitude, longitude], radius=2000, fill=False, color='white').add_to(map_atl)
folium.Circle([latitude, longitude], radius=3000, fill=False, color='white').add_to(map_atl)
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

Looks like a few pockets of low restaurant density closest to city center can be found **north west ,south west and south east of Atlanta center**. 

Let's create another heatmap map showing **heatmap/density of Vietnamese restaurants** only.

In [46]:
map_atl = folium.Map(location=[latitude, longitude], zoom_start=13)
folium.TileLayer('cartodbpositron').add_to(map_atl) #cartodbpositron cartodbdark_matter
HeatMap(vietnamese_latlons).add_to(map_atl)
folium.Marker([latitude, longitude]).add_to(map_atl)
folium.Circle([latitude, longitude], radius=1000, fill=False, color='white').add_to(map_atl)
folium.Circle([latitude, longitude], radius=2000, fill=False, color='white').add_to(map_atl)
folium.Circle([latitude, longitude], radius=3000, fill=False, color='white').add_to(map_atl)
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from unicodedata import normalize

from bs4 import BeautifulSoup
    
url1 = "http://zipatlas.com/us/ga/atlanta/zip-code-comparison/population-density.htm"
page = requests.get(url1)    
soup = BeautifulSoup(page.content, 'lxml')
html_string = '''
<tr><td><div style="padding: 0px 0px 15px 0px;">Related Reports<div style="height:3px;"></div><div style="padding-left:15px;"><a href="/us/ga/zip-code-comparison/population-density.htm">Population Density in Georgia by Zip Code</a></div><div style="padding-left:15px;"><a href="/us/ga/city-comparison/population-density.htm">Population Density in Georgia by City</a></div><div style="padding-left:15px;"><a href="/us/zip-code-comparison/population-density.htm">Population Density in the United States by Zip Code</a></div><div style="padding-left:15px;"><a href="/us/city-comparison/population-density.htm">Population Density in the United States by City</a></div></div><table rules="all" frame="box" style="width:100%;"><tr><td class="report_header" align="center" style="width:0%;">#</td><td class="report_header" align="center" style="width:8%;">Zip Code</td><td class="report_header" align="center" style="width:19%;">Location</td><td class="report_header" align="center" style="width:25%;">City</td><td class="report_header" align="center" style="width:13%;">Population</td><td class="report_header" align="center" style="width:22%;">People / Sq. Mile</td><td class="report_header" align="center" style="width:13%;">National Rank</td></tr><tr><td class="report_data" align="right">1.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30313.htm">30313</a></td><td class="report_data" align="right">33.764661, -84.397547</td><td class="report_data" align="left"><a href="/us/ga/atlanta.htm">Atlanta</a>, <a href="/us/georgia.htm">Georgia</a></td><td class="report_data" align="right">11,035</td><td class="report_data" align="right"><b>9,768.73</b></td><td class="report_data" align="right">#617</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">2.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30322.htm">30322</a></td><td class="report_data" align="right">33.795590, -84.327312</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">1,724</td><td class="report_data" align="right"><b>8,794.33</b></td><td class="report_data" align="right">#712</td></tr><tr><td class="report_data" align="right">3.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30308.htm">30308</a></td><td class="report_data" align="right">33.770877, -84.377873</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">11,796</td><td class="report_data" align="right"><b>7,377.75</b></td><td class="report_data" align="right">#947</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">4.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30312.htm">30312</a></td><td class="report_data" align="right">33.745377, -84.376180</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">20,221</td><td class="report_data" align="right"><b>6,289.52</b></td><td class="report_data" align="right">#1,230</td></tr><tr><td class="report_data" align="right">5.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30314.htm">30314</a></td><td class="report_data" align="right">33.757777, -84.431884</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">27,181</td><td class="report_data" align="right"><b>5,774.91</b></td><td class="report_data" align="right">#1,393</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">6.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30309.htm">30309</a></td><td class="report_data" align="right">33.799832, -84.385903</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">18,739</td><td class="report_data" align="right"><b>5,312.42</b></td><td class="report_data" align="right">#1,577</td></tr><tr><td class="report_data" align="right">7.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30329.htm">30329</a></td><td class="report_data" align="right">33.827479, -84.323017</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">27,872</td><td class="report_data" align="right"><b>4,904.58</b></td><td class="report_data" align="right">#1,762</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">8.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30306.htm">30306</a></td><td class="report_data" align="right">33.788547, -84.349985</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">21,199</td><td class="report_data" align="right"><b>4,888.18</b></td><td class="report_data" align="right">#1,771</td></tr><tr><td class="report_data" align="right">9.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30324.htm">30324</a></td><td class="report_data" align="right">33.818405, -84.358175</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">20,317</td><td class="report_data" align="right"><b>3,852.84</b></td><td class="report_data" align="right">#2,435</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">10.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30310.htm">30310</a></td><td class="report_data" align="right">33.726586, -84.425995</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">33,598</td><td class="report_data" align="right"><b>3,817.23</b></td><td class="report_data" align="right">#2,467</td></tr><tr><td class="report_data" align="right">11.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30317.htm">30317</a></td><td class="report_data" align="right">33.748355, -84.317585</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">13,981</td><td class="report_data" align="right"><b>3,810.85</b></td><td class="report_data" align="right">#2,469</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">12.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30319.htm">30319</a></td><td class="report_data" align="right">33.877450, -84.336316</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">36,783</td><td class="report_data" align="right"><b>3,689.26</b></td><td class="report_data" align="right">#2,565</td></tr><tr><td class="report_data" align="right">13.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30307.htm">30307</a></td><td class="report_data" align="right">33.771079, -84.333600</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">16,646</td><td class="report_data" align="right"><b>3,624.86</b></td><td class="report_data" align="right">#2,632</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">14.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30303.htm">30303</a></td><td class="report_data" align="right">33.752717, -84.390679</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">3,744</td><td class="report_data" align="right"><b>3,530.81</b></td><td class="report_data" align="right">#2,723</td></tr><tr><td class="report_data" align="right">15.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30315.htm">30315</a></td><td class="report_data" align="right">33.702829, -84.382515</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">40,288</td><td class="report_data" align="right"><b>3,502.52</b></td><td class="report_data" align="right">#2,746</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">16.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30344.htm">30344</a></td><td class="report_data" align="right">33.676198, -84.458359</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">36,753</td><td class="report_data" align="right"><b>3,304.67</b></td><td class="report_data" align="right">#2,909</td></tr><tr><td class="report_data" align="right">17.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30340.htm">30340</a></td><td class="report_data" align="right">33.898619, -84.249569</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">26,997</td><td class="report_data" align="right"><b>3,173.45</b></td><td class="report_data" align="right">#3,036</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">18.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30305.htm">30305</a></td><td class="report_data" align="right">33.836781, -84.391308</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">21,380</td><td class="report_data" align="right"><b>3,157.98</b></td><td class="report_data" align="right">#3,047</td></tr><tr><td class="report_data" align="right">19.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30345.htm">30345</a></td><td class="report_data" align="right">33.852060, -84.281753</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">22,791</td><td class="report_data" align="right"><b>3,078.99</b></td><td class="report_data" align="right">#3,123</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">20.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30342.htm">30342</a></td><td class="report_data" align="right">33.884350, -84.374964</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">25,807</td><td class="report_data" align="right"><b>2,922.81</b></td><td class="report_data" align="right">#3,293</td></tr><tr><td class="report_data" align="right">21.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30341.htm">30341</a></td><td class="report_data" align="right">33.887723, -84.291143</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">31,036</td><td class="report_data" align="right"><b>2,898.57</b></td><td class="report_data" align="right">#3,315</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">22.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30338.htm">30338</a></td><td class="report_data" align="right">33.946218, -84.318119</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">28,885</td><td class="report_data" align="right"><b>2,878.51</b></td><td class="report_data" align="right">#3,340</td></tr><tr><td class="report_data" align="right">23.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30360.htm">30360</a></td><td class="report_data" align="right">33.933808, -84.272577</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">14,782</td><td class="report_data" align="right"><b>2,873.23</b></td><td class="report_data" align="right">#3,353</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">24.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30311.htm">30311</a></td><td class="report_data" align="right">33.723262, -84.475994</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">35,230</td><td class="report_data" align="right"><b>2,787.45</b></td><td class="report_data" align="right">#3,448</td></tr><tr><td class="report_data" align="right">25.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30316.htm">30316</a></td><td class="report_data" align="right">33.711546, -84.331796</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">32,495</td><td class="report_data" align="right"><b>2,593.76</b></td><td class="report_data" align="right">#3,661</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">26.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30318.htm">30318</a></td><td class="report_data" align="right">33.791431, -84.446141</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">53,229</td><td class="report_data" align="right"><b>2,548.50</b></td><td class="report_data" align="right">#3,711</td></tr><tr><td class="report_data" align="right">27.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30350.htm">30350</a></td><td class="report_data" align="right">33.977483, -84.332407</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">31,057</td><td class="report_data" align="right"><b>2,487.16</b></td><td class="report_data" align="right">#3,783</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">28.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30339.htm">30339</a></td><td class="report_data" align="right">33.872067, -84.462985</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">15,782</td><td class="report_data" align="right"><b>2,240.62</b></td><td class="report_data" align="right">#4,106</td></tr><tr><td class="report_data" align="right">29.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30328.htm">30328</a></td><td class="report_data" align="right">33.930598, -84.384238</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">28,701</td><td class="report_data" align="right"><b>2,198.56</b></td><td class="report_data" align="right">#4,155</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">30.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30354.htm">30354</a></td><td class="report_data" align="right">33.660361, -84.386502</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">17,927</td><td class="report_data" align="right"><b>1,639.24</b></td><td class="report_data" align="right">#5,011</td></tr><tr><td class="report_data" align="right">31.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30326.htm">30326</a></td><td class="report_data" align="right">33.849787, -84.364521</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">1,075</td><td class="report_data" align="right"><b>1,425.22</b></td><td class="report_data" align="right">#5,379</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">32.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30337.htm">30337</a></td><td class="report_data" align="right">33.640147, -84.450081</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">16,084</td><td class="report_data" align="right"><b>1,370.83</b></td><td class="report_data" align="right">#5,495</td></tr><tr><td class="report_data" align="right">33.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30346.htm">30346</a></td><td class="report_data" align="right">33.925077, -84.338507</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">1,452</td><td class="report_data" align="right"><b>1,344.19</b></td><td class="report_data" align="right">#5,544</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">34.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30327.htm">30327</a></td><td class="report_data" align="right">33.869768, -84.417135</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">21,003</td><td class="report_data" align="right"><b>1,183.09</b></td><td class="report_data" align="right">#5,908</td></tr><tr><td class="report_data" align="right">35.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30349.htm">30349</a></td><td class="report_data" align="right">33.626684, -84.523669</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">55,239</td><td class="report_data" align="right"><b>1,163.96</b></td><td class="report_data" align="right">#5,949</td></tr><tr style="background-color:#f0f5f9;"><td class="report_data" align="right">36.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30331.htm">30331</a></td><td class="report_data" align="right">33.708174, -84.542280</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">44,444</td><td class="report_data" align="right"><b>1,133.56</b></td><td class="report_data" align="right">#6,009</td></tr><tr><td class="report_data" align="right">37.</td><td class="report_data" align="center"><a href="/us/ga/atlanta/zip-30336.htm">30336</a></td><td class="report_data" align="right">33.735413, -84.568933</td><td class="report_data" align="left"><span style="color:#7070A0;" class="link" onclick="window.location='/us/ga/atlanta.htm';">Atlanta</span>, <span style="color:#7070A0;" class="link" onclick="window.location='/us/georgia.htm';">Georgia</span></td><td class="report_data" align="right">238</td><td class="report_data" align="right"><b>27.21</b></td><td class="report_data" align="right">#22,215</td></tr></table></td></tr>
'''
    
df=pd.read_html(html_string)[0]
new_header = df.iloc[0] 
df = df[1:] 
df.columns = new_header 
df=df.drop(['#','City','Population','National Rank'],axis=1)
 
df.head(5)

Let's take a look at the population density within Atlanta city

In [51]:
df_stat=pd.read_csv('C:/Users/daile/Dropbox/R Python/City_of_Atlanta_Neighborhood_Statistical_Areas (1).csv')
df_stat=df_stat.drop(['OBJECTID','NPU','STATISTICA','URL','last_edited_date'],axis=1)
# Sort Neighborhood according to population in descending order
df_stat=df_stat.sort_values(by="pop",ascending=False)
df_stat["NEIGHBORHO"]=df_stat["NEIGHBORHO"].astype(str)
df_stat["pop"]=df_stat["pop"].astype(float)
df_stat=df_stat.sort_values(by="asian",ascending=False)
df_stat=df_stat.drop(["POP2010","A","GlobalID"],axis=1)
df_stat

In [52]:
!pip install altair
!pip install altair_saver
import altair as alt
import selenium.webdriver
driver = selenium.webdriver.Chrome()
chart=alt.Chart(df_stat).mark_bar(
    size=50
).encode(
    alt.X("NEIGHBORHO", title="Neighborhood", scale=alt.Scale(zero=False),type='nominal'),
    alt.Y("pop", title="Population", scale=alt.Scale(zero=False)),
    tooltip=["NEIGHBORHO", "pop"]
).properties(
    height=300,
    width=1500,
).interactive()
chart

ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'c:\\users\\daile\\anaconda3\\lib\\site-packages\\pandas-1.1.4.dist-info\\METADATA'

ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: 'c:\\users\\daile\\anaconda3\\lib\\site-packages\\pandas-1.1.4.dist-info\\METADATA'



alt.Chart(...)

In [ ]:
chart.save('Atl_pop.png', scale_factor=4.0)

Our candidates are neighborhoods with high asian population, which are Home Park and Georgia Tech. Let's use google map to explore briefly these neighborhoods

In [ ]:
import requests, json 
  
# enter your api key here 
api_key = 'AIzaSyDhYYiEvsDxhKeU9jXM69XniQhNct1x5zU'
  
# url variable store url 
url = "https://maps.googleapis.com/maps/api/place/textsearch/json?"
  
# The text string on which to search 
query = input('Search query: ') 
  
# get method of requests module 
# return response object 
r = requests.get(url + 'query=' + query +
                        '&key=' + api_key) 
  
# json method of response object convert 
#  json format data into python format data 
x = r.json() 
  
# now x contains list of nested dictionaries 
# we know dictionary contain key value pair 
# store the value of result key in variable y 
y = x['results'] 
  
# keep looping upto length of y 
for i in range(len(y)): 
      
    # Print value corresponding to the 
    # 'name' key at the ith index of y 
    print(y[i]['name']) 

Let's define new, more narrow region of interest, which will include low-restaurant-count parts of Georgia Tech, Home Park and Atlantic Station. Since they are close to each other, let's take a look at another location further away where restaurant density is low, which a Grant Park

In [53]:
roi_x_min = atl_center_x - 2000
roi_y_max = atl_center_y + 1000
roi_width = 5000
roi_height = 5000
roi_center_x = roi_x_min + 2500
roi_center_y = roi_y_max - 2500
roi_center_lon, roi_center_lat = xy_to_lonlat(roi_center_x, roi_center_y)
roi_center = [roi_center_lat, roi_center_lon]

map_atl = folium.Map(location=roi_center, zoom_start=14)
HeatMap(restaurant_latlons).add_to(map_atl)
folium.Marker([latitude, longitude]).add_to(map_atl)
folium.Circle(roi_center, radius=2500, color='white', fill=True, fill_opacity=0.4).add_to(map_atl)
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

Let's also create new, more dense grid of location candidates restricted to our new region of interest (let's make our location candidates 100m appart).

In [54]:
k = math.sqrt(3) / 2 # Vertical offset for hexagonal grid cells
x_step = 100
y_step = 100 * k 
roi_y_min = roi_center_y - 2500

roi_latitudes = []
roi_longitudes = []
roi_xs = []
roi_ys = []
for i in range(0, int(51/k)):
    y = roi_y_min + i * y_step
    x_offset = 50 if i%2==0 else 0
    for j in range(0, 51):
        x = roi_x_min + j * x_step + x_offset
        d = calc_xy_distance(roi_center_x, roi_center_y, x, y)
        if (d <= 2501):
            lon, lat = xy_to_lonlat(x, y)
            roi_latitudes.append(lat)
            roi_longitudes.append(lon)
            roi_xs.append(x)
            roi_ys.append(y)

print(len(roi_latitudes), 'candidate neighborhood centers generated.')

2261 candidate neighborhood centers generated.


2261 candidate neighborhood centers generated
OK. Now let's calculate two most important things for each location candidate: number of restaurants in vicinity (we'll use radius of 250 meters) and distance to closest Vietnamese restaurant.

In [55]:
def count_restaurants_nearby(x, y, restaurants, radius=250):    
    count = 0
    for res in restaurants.values():
        res_x = res[7]; res_y = res[8]
        d = calc_xy_distance(x, y, res_x, res_y)
        if d<=radius:
            count += 1
    return count

def find_nearest_restaurant(x, y, restaurants):
    d_min = 100000
    for res in restaurants.values():
        res_x = res[7]; res_y = res[8]
        d = calc_xy_distance(x, y, res_x, res_y)
        if d<=d_min:
            d_min = d
    return d_min

roi_restaurant_counts = []
roi_vietnamese_distances = []

print('Generating data on location candidates... ', end='')
for x, y in zip(roi_xs, roi_ys):
    count = count_restaurants_nearby(x, y, restaurants, radius=250)
    roi_restaurant_counts.append(count)
    distance = find_nearest_restaurant(x, y, vietnamese_restaurants)
    roi_vietnamese_distances.append(distance)
print('done.')


Generating data on location candidates... done.


In [58]:
# Let's put this into dataframe
df_roi_locations = pd.DataFrame({'Latitude':roi_latitudes,
                                 'Longitude':roi_longitudes,
                                 'X':roi_xs,
                                 'Y':roi_ys,
                                 'Restaurants nearby':roi_restaurant_counts,
                                 'Distance to Vietnamese restaurant':roi_vietnamese_distances})

df_roi_locations.head(10)

Latitude  Longitude             X             Y  Restaurants nearby  \
0  33.757186 -84.367184 -6.878425e+06  1.151568e+07                   1   
1  33.757683 -84.367359 -6.878325e+06  1.151568e+07                   0   
2  33.754327 -84.366737 -6.878975e+06  1.151577e+07                   1   
3  33.754824 -84.366912 -6.878875e+06  1.151577e+07                   1   
4  33.755320 -84.367087 -6.878775e+06  1.151577e+07                   2   
5  33.755817 -84.367262 -6.878675e+06  1.151577e+07                   0   
6  33.756314 -84.367437 -6.878575e+06  1.151577e+07                   0   
7  33.756811 -84.367612 -6.878475e+06  1.151577e+07                   0   
8  33.757308 -84.367787 -6.878375e+06  1.151577e+07                   0   
9  33.757805 -84.367962 -6.878275e+06  1.151577e+07                   0   

   Distance to Vietnamese restaurant  
0                         515.225411  
1                         550.890078  
2                         712.186124  
3                         659.720948  
4                         619.075427  
5                         592.686290  
6                         582.494372  
7                         589.340604  
8                         612.654064  
9                         650.667086

OK. Let us now filter those locations: we're interested only in locations with no more than two restaurants in radius of 250 meters, and no Vietnamese restaurants in radius of 400 meters.

In [59]:
good_res_count = np.array((df_roi_locations['Restaurants nearby']<=2))
print('Locations with no more than two restaurants nearby:', good_res_count.sum())

good_vn_distance = np.array(df_roi_locations['Distance to Vietnamese restaurant']>=400)
print('Locations with no Vietnamese restaurants within 400m:', good_vn_distance.sum())

good_locations = np.logical_and(good_res_count, good_vn_distance)
print('Locations with both conditions met:', good_locations.sum())

df_good_locations = df_roi_locations[good_locations]


Locations with no more than two restaurants nearby: 1882
Locations with no Vietnamese restaurants within 400m: 1964
Locations with both conditions met: 1742


Let's see how this looks on a map.

In [60]:
good_latitudes = df_good_locations['Latitude'].values
good_longitudes = df_good_locations['Longitude'].values

good_locations = [[lat, lon] for lat, lon in zip(good_latitudes, good_longitudes)]

map_atl = folium.Map(location=roi_center, zoom_start=14)
folium.TileLayer('cartodbpositron').add_to(map_atl)
HeatMap(restaurant_latlons).add_to(map_atl)
folium.Circle(roi_center, radius=2500, color='white', fill=True, fill_opacity=0.6).add_to(map_atl)
folium.Marker([latitude, longitude]).add_to(map_atl)
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_atl) 
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

We now have a bunch of locations fairly close to downtown  and we know that each of those locations has no more than two restaurants in radius of 250m, and no Vietnamese restaurant closer than 400m. Any of those locations is a potential candidate for a new Vietnamese restaurant, at least based on nearby competition. However, it rules out the area around GA Tech. So we only have Grant Park left

Let's now show those good locations in a form of heatmap:

In [61]:
map_atl = folium.Map(location=roi_center, zoom_start=14)
HeatMap(good_locations, radius=25).add_to(map_atl)
folium.Marker([latitude, longitude]).add_to(map_atl)
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_atl)
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

Looking good. What we have now is a clear indication of zones with low number of restaurants in vicinity, and no Vietnamese restaurants at all nearby.

Let us now cluster those locations to create centers of zones containing good locations. Those zones, their centers and addresses will be the final result of our analysis.

In [63]:
from sklearn.cluster import KMeans

number_of_clusters = 15

good_xys = df_good_locations[['X', 'Y']].values
kmeans = KMeans(n_clusters=number_of_clusters, random_state=0).fit(good_xys)

cluster_centers = [xy_to_lonlat(cc[0], cc[1]) for cc in kmeans.cluster_centers_]

map_atl = folium.Map(location=roi_center, zoom_start=14)
folium.TileLayer('cartodbpositron').add_to(map_atl)
HeatMap(restaurant_latlons).add_to(map_atl)
folium.Circle(roi_center, radius=2500, color='white', fill=True, fill_opacity=0.4).add_to(map_atl)
folium.Marker([latitude, longitude]).add_to(map_atl)
for lon, lat in cluster_centers:
    folium.Circle([lat, lon], radius=500, color='green', fill=True, fill_opacity=0.25).add_to(map_atl) 
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_atl)
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

Not bad - our clusters represent groupings of most of the candidate locations and cluster centers are placed nicely in the middle of the zones 'rich' with location candidates.

Addresses of those cluster centers will be a good starting point for exploring the neighborhoods to find the best possible location based on neighborhood specifics.

Let's see those zones on a city map without heatmap, using shaded areas to indicate our clusters:

In [ ]:
map_atl = folium.Map(location=roi_center, zoom_start=14)
folium.Marker([latitude, longitude]).add_to(map_atl)
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.Circle([lat, lon], radius=250, color='#00000000', fill=True, fill_color='#0066ff', fill_opacity=0.07).add_to(map_atl)
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_atl)
for lon, lat in cluster_centers:
    folium.Circle([lat, lon], radius=500, color='green', fill=False).add_to(map_atl) 
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

Let's zoom in on candidate areas in Grant Park

In [ ]:
map_atl= folium.Map(location=[33.74095, -84.369977], zoom_start=17)
folium.Marker([latitude,longitude]).add_to(map_atl)
for lon, lat in cluster_centers:
    folium.Circle([lat, lon], radius=500, color='green', fill=False).add_to(map_atl) 
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.Circle([lat, lon], radius=250, color='#0000ff00', fill=True, fill_color='#0066ff', fill_opacity=0.07).add_to(map_atl)
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_atl)
folium.GeoJson(atlanta_data, style_function=boroughs_style, name='geojson').add_to(map_atl)
map_atl

Finaly, let's reverse geocode those candidate area centers to get the addresses which can be presented to stakeholders.

In [64]:
candidate_area_addresses = []
print('==============================================================')
print('Addresses of centers of areas recommended for further analysis')
print('==============================================================\n')
for lon, lat in cluster_centers:
    addr = get_address(google_api_key, lat, lon).replace(', USA', '')
    candidate_area_addresses.append(addr)    
    x, y = lonlat_to_xy(lon, lat)
    d = calc_xy_distance(x, y, atl_center_x, atl_center_y)
    print('{}{} => {:.1f}km from Downtown'.format(addr, ' '*(50-len(addr)), d/1000))
    

Addresses of centers of areas recommended for further analysis

227 Centennial Olympic Park Dr NW, Atlanta, GA 30303 => 2.3km from Downtown
400 Decatur St SE, Atlanta, GA 30312               => 2.4km from Downtown
7 M.L.K. Jr Dr SW, Atlanta, GA 30303               => 0.5km from Downtown
188 Boulevard NE, Atlanta, GA 30312                => 3.5km from Downtown
125 Logan St SE, Atlanta, GA 30312                 => 1.8km from Downtown
44 Courtland St NE, Atlanta, GA 30303              => 1.7km from Downtown
Downtown Connector, Atlanta, GA 30303              => 2.8km from Downtown
170 Boulevard SE, Atlanta, GA 30312                => 3.2km from Downtown
263 Decatur St SE, Atlanta, GA 30312               => 1.3km from Downtown
3152 M.L.K. Jr Dr NW, Atlanta, GA 30303            => 1.1km from Downtown
18 Howell St NE, Atlanta, GA 30312                 => 3.4km from Downtown
357 Oakland Park SE, Atlanta, GA 30312             => 2.5km from Downtown
390 Piedmont Ave NE, Atlanta, GA 30308        

This concludes our analysis. We have created 15 addresses representing centers of zones containing locations with low number of restaurants and no Vietnamese restaurants nearby, all zones being fairly close to city center (all less than 4km from Downtown, and about half of those less than 2km from Downtown). Although zones are shown on map with a radius of ~500 meters (green circles), their shape is actually very irregular and their centers/addresses should be considered only as a starting point for exploring area neighborhoods in search for potential restaurant locations. These zones are close to Zoo Atlanta and Emory University which make them fairly attractive. Other areas close to GA Tech and Atlantic Station are saturated with restaurants.


In [67]:
map_atl= folium.Map(location=roi_center, zoom_start=14)
folium.Circle([latitude,longitude], radius=50, color='red', fill=True, fill_color='red', fill_opacity=1).add_to(map_atl)
for lonlat, addr in zip(cluster_centers, candidate_area_addresses):
    folium.Marker([lonlat[1], lonlat[0]], popup=addr).add_to(map_atl) 
for lat, lon in zip(good_latitudes, good_longitudes):
    folium.Circle([lat, lon], radius=250, color='#0000ff00', fill=True, fill_color='#0066ff', fill_opacity=0.05).add_to(map_atl)
map_atl

## Results and Discussion <a name="results"></a>

Our analysis shows that although there is a great number of restaurants in Atlanta (~2000 in our initial area of interest which was 12x12km around Downtown), there are pockets of low restaurant density fairly close to city center. Highest concentration of restaurants was detected north and west from Downtown, so we focused our attention to areas south, south-east and east, corresponding to Grant Park, Peachtree center, Five Points. They are close to Zoo Atlanta, Emory University. 

After directing our attention to this narrower area of interest (covering approx. 5x5km south-east from Downtown) we first created a dense grid of location candidates (spaced 100m apart); those locations were then filtered so that those with more than two restaurants in radius of 250m and those with an Vietnamese restaurant closer than 400m were removed.

Those location candidates were then clustered to create zones of interest which contain greatest number of location candidates. Addresses of centers of those zones were also generated using reverse geocoding to be used as markers/starting points for more detailed local analysis based on other factors.

Result of all this is 15 zones containing largest number of potential new restaurant locations based on number of and distance to existing venues - both restaurants in general and Vietnamese restaurants particularly. This, of course, does not imply that those zones are actually optimal locations for a new restaurant! Purpose of this analysis was to only provide info on areas close to Atlanta center but not crowded with existing restaurants (particularly Vietnamese) - it is entirely possible that there is a very good reason for small number of restaurants in any of those areas, reasons which would make them unsuitable for a new restaurant regardless of lack of competition in the area. Recommended zones should therefore be considered only as a starting point for more detailed analysis which could eventually result in location which has not only no nearby competition but also other factors taken into account and all other relevant conditions met.

## Conclusion <a name="conclusion"></a>

Purpose of this project was to identify Atlanta areas close to center with low number of restaurants (particularly Vietnamese restaurants) in order to aid stakeholders in narrowing down the search for optimal location for a new Vietnamese restaurant. By calculating restaurant density distribution from Foursquare data we have first identified general boroughs that justify further analysis (Grant Park, GA Tech), and then generated extensive collection of locations which satisfy some basic requirements regarding existing nearby restaurants. Clustering of those locations was then performed in order to create major zones of interest (containing greatest number of potential locations) and addresses of those zone centers were created to be used as starting points for final exploration by stakeholders.

Final decision on optimal restaurant location will be made by stakeholders based on specific characteristics of neighborhoods and locations in every recommended zone, taking into consideration additional factors like attractiveness of each location (proximity to park or water), levels of noise / proximity to major roads, real estate availability, prices, social and economic dynamics of every neighborhood etc.